syz-ci/jobs.go

```diff
@@ -26,72 +27,146 @@ import (
 	"github.com/google/syzkaller/vm"
 )
 
+type JobManager struct {
+	cfg             *Config
+	dash            *dashapi.Dashboard
+	managers        []*Manager
+	shutdownPending <-chan struct{}
+}
+
 type JobProcessor struct {
 	cfg             *Config
 	name            string
+	instanceSuffix  string
 	managers        []*Manager
 	knownCommits    map[string]bool
-	stop            chan struct{}
-	shutdownPending chan struct{}
 	dash            *dashapi.Dashboard
-	syzkallerRepo   string
-	syzkallerBranch string
+	baseDir         string
+	jobFilter       *ManagerJobs
+	jobTicker       <-chan time.Time
+	commitTicker    <-chan time.Time
+	shutdownPending <-chan struct{}
 }
 
-func newJobProcessor(cfg *Config, managers []*Manager, stop, shutdownPending chan struct{}) (*JobProcessor, error) {
+func newJobManager(cfg *Config, managers []*Manager, shutdownPending chan struct{}) (*JobManager, error) {
 	dash, err := dashapi.New(cfg.DashboardClient, cfg.DashboardAddr, cfg.DashboardKey)
 	if err != nil {
 		return nil, err
 	}
-	return &JobProcessor{
+	return &JobManager{
 		cfg:             cfg,
-		name:            fmt.Sprintf("%v-job", cfg.Name),
+		dash:            dash,
 		managers:        managers,
-		knownCommits:    make(map[string]bool),
-		stop:            stop,
 		shutdownPending: shutdownPending,
-		dash:            dash,
-		syzkallerRepo:   cfg.SyzkallerRepo,
-		syzkallerBranch: cfg.SyzkallerBranch,
 	}, nil
 }
 
-func (jp *JobProcessor) loop() {
-	jobTicker := time.NewTicker(time.Duration(jp.cfg.JobPollPeriod) * time.Second)
-	commitTicker := time.NewTicker(time.Duration(jp.cfg.CommitPollPeriod) * time.Second)
-	defer jobTicker.Stop()
+// backgroundLoop starts a job loop in parallel and returns a blocking function
+// to gracefully stop job processing.
+func (jm *JobManager) backgroundLoop(wg *sync.WaitGroup) func() {
+	stop := make(chan struct{})
+	done := make(chan struct{}, 1)
+	go func() {
+		jm.loop(stop)
+		done <- struct{}{}
```

Comment:
This done channel looks excessive, there is already a separate WaitGroup used in the loop and it's signalled exactly when both loops exit. It looks simple to just wait for that WaitGroup.

Reply:
I fear it could get messy, because in `startLoop` we already deal with one WaitGroup -- the global one. Adding one more would make the code much less readable, a channel seemed a harmless alternative.

If there were some way to to create some "sub group" of the global wait group, then I could just instantiate it in `startLoop` and pass to the `loop` function. But AFAIK there's no such thing in the standard Go library.